pkg/trace/writer/trace.go

```diff
@@ -127,183 +120,202 @@ func NewTraceWriter(
+	w.stats.Traces.Add(int64(len(pkg.TracerPayload.Chunks)))
+	w.stats.Events.Add(pkg.EventCount)
+
+	toflush := w.appendChunks(pkg)
+	if toflush != nil {
+		w.flushPayloads(toflush)
 	}
-	w.flush()
-	w.swg.Wait()
 }
-
 func (w *TraceWriter) resetBuffer() {
 	w.bufferedSize = 0
 	w.tracerPayloads = make([]*pb.TracerPayload, 0, len(w.tracerPayloads))
 }
 
 const headerLanguages = "X-Datadog-Reported-Languages"
 
+// w must be locked for a flush.
 func (w *TraceWriter) flush() {
-	if len(w.tracerPayloads) == 0 {
+	defer w.resetBuffer()
+	w.flushPayloads(w.tracerPayloads)
+}
+
+var gzpool = sync.Pool{}
+
+func getGZW(w io.Writer) (*gzip.Writer, error) {
+	gw := gzpool.Get()
+	if gw == nil {
+		return gzip.NewWriterLevel(w, gzip.BestSpeed)
+	}
+	gws := gw.(*gzip.Writer)
+	gws.Reset(w)
+	return gws, nil
+}
+
+// w does not need to be locked during flushPayloads.
+func (w *TraceWriter) flushPayloads(payloads []*pb.TracerPayload) {
+	w.flushTicker.Reset(w.tick) // reset the flush timer whenever we flush
+	if len(payloads) == 0 {
 		// nothing to do
 		return
 	}
 
 	defer w.timing.Since("datadog.trace_agent.trace_writer.encode_ms", time.Now())
-	defer w.resetBuffer()
 
-	log.Debugf("Serializing %d tracer payloads.", len(w.tracerPayloads))
+	log.Debugf("Serializing %d tracer payloads.", len(payloads))
 	p := pb.AgentPayload{
 		AgentVersion:       w.agentVersion,
 		HostName:           w.hostname,
 		Env:                w.env,
 		TargetTPS:          w.prioritySampler.GetTargetTPS(),
 		ErrorTPS:           w.errorsSampler.GetTargetTPS(),
 		RareSamplerEnabled: w.rareSampler.IsEnabled(),
-		TracerPayloads:     w.tracerPayloads,
+		TracerPayloads:     payloads,
 	}
 	log.Debugf("Reported agent rates: target_tps=%v errors_tps=%v rare_sampling=%v", p.TargetTPS, p.ErrorTPS, p.RareSamplerEnabled)
 
-	w.swg.Add(1)
-	w.Serialize <- &p
+	w.serialize(&p)
 }
 
-func (w *TraceWriter) serializer() {
-	defer w.wg.Done()
-	for pl := range w.Serialize {
-		func() {
-			defer w.swg.Done()
-			b, err := pl.MarshalVT()
-			if err != nil {
-				log.Errorf("Failed to serialize payload, data dropped: %v", err)
-				return
-			}
-
-			w.stats.BytesUncompressed.Add(int64(len(b)))
-			p := newPayload(map[string]string{
-				"Content-Type":     "application/x-protobuf",
-				"Content-Encoding": "gzip",
-				headerLanguages:    strings.Join(info.Languages(), "|"),
-			})
-			p.body.Grow(len(b) / 2)
-			gzipw, err := gzip.NewWriterLevel(p.body, gzip.BestSpeed)
-			if err != nil {
-				// it will never happen, unless an invalid compression is chosen;
-				// we know gzip.BestSpeed is valid.
-				log.Errorf("gzip.NewWriterLevel: %d", err)
-				return
-			}
-			if _, err := gzipw.Write(b); err != nil {
-				log.Errorf("Error gzipping trace payload: %v", err)
-			}
-			if err := gzipw.Close(); err != nil {
-				log.Errorf("Error closing gzip stream when writing trace payload: %v", err)
-			}
-			sendPayloads(w.senders, p, w.syncMode)
-		}()
+var outPool = sync.Pool{}
```

Comment:
can we add a small explanation on why we are using this pool here? It is clear in the context of this PR, but we should maybe drop a line explaining the rationale for the future 🙏